File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.cpp
class Solution {
public:
    string minRemoveToMakeValid(string s) {
        stack<int> par;
        for (int i = 0; i < s.length(); i++) {
            if (s[i] == '(') {
                par.push(i);
            } else if (s[i] == ')') {
                if (!par.empty()) {
                    par.pop();
                } else {
                    s[i] = '*'; // Marking ')' for removal
                }
            }
        }

        while (!par.empty()) {
            s[par.top()] = '*'; // Marking '(' for removal
            par.pop();
        }

        string result;
        for (char c : s) {
            if (c != '*') {
                result += c;
            }
        }

        return result;
    }
};

File: 1249-minimum-remove-to-make-valid-parentheses/1249-minimum-remove-to-make-valid-parentheses.py
class Solution:
    def minRemoveToMakeValid(self, s: str) -> str:
        par = []
        for i in range(len(s)):
            if s[i] == '(':
                par.append(i)
            elif s[i] == ')':
                if par:
                    par.pop()
                else:
                    s = s[:i] + '*' + s[i+1:]  # Marking ')' for removal
        
        while par:
            s = s[:par[-1]] + '*' + s[par[-1]+1:]  # Marking '(' for removal
            par.pop()
        
        result = ''
        for c in s:
            if c != '*':
                result += c
        
        return result